internal/grpc/grpc.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+package grpc
+
+import (
+	"crypto/tls"
+	"errors"
+
+	grpcRetry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"google.golang.org/grpc/credentials/local"
+
+	"github.com/nginx/agent/v3/internal/config"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+// these will come from the agent config
+var serviceConfig = `{
+	"loadBalancingPolicy": "round_robin",
+	"healthCheckConfig": {
+		"serviceName": "nginx-agent"
+	}
+}`
+var defaultCredentials = local.NewCredentials()
+
+func GetDialOptions(agentConfig *config.Config) []grpc.DialOption {
+	opts := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithReturnConnectionError(),
+		grpc.WithStreamInterceptor(grpcRetry.StreamClientInterceptor()),
+		grpc.WithUnaryInterceptor(grpcRetry.UnaryClientInterceptor()),
+		grpc.WithDefaultServiceConfig(serviceConfig),
+	}
+
+	if agentConfig.Client != nil {
+		keepAlive := keepalive.ClientParameters{
+			Time:                agentConfig.Client.Time,
+			Timeout:             agentConfig.Client.Timeout,
+			PermitWithoutStream: agentConfig.Client.PermitStream,
```

Comment:
Is there a risk of the user setting the wrong boolean value for `agentConfig.Client.PermitStream` considering that the actual name of the keepalive parameter is `PermitWithoutStream`?

Reply:
renamed to be more appropriate, see next commit